File: Java-Civitas/MazoSorpresas.java
package civitas;

import java.util.ArrayList;
import java.util.Collections;
        
public class MazoSorpresas {
    
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init() {
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        usadas = 0;
        barajada = false;
    }
    
    public MazoSorpresas() {
        init();
        debug = false;
    }
    
    public MazoSorpresas(boolean d) {
        debug = d;
        init(); 
        Diario.getInstance().ocurreEvento("Debug Mode:" + debug);
    }
    
    void alMazo(Sorpresa s) {
        if(!barajada)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente() {
        if((!barajada || usadas == sorpresas.size()) && !debug){
            Collections.shuffle(sorpresas);
            barajada = true;
            usadas = 0;
            usadas++;
            Sorpresa temp = sorpresas.get(0);
            sorpresas.remove(0);
            sorpresas.add(temp);
            ultimaSorpresa = temp;
        }
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa) {
        if(sorpresas.contains(sorpresa)){
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento(sorpresa + "Carta aniadida a cartas especiales");
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa) {
        if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
             Diario.getInstance().ocurreEvento(sorpresa + "Carta aniadida a sorpresas");
        }
    }
}

File: Ruby-Civitas/mazo_sorpresas.rb
module Civitas
  class MazoSorpresas
    
    def initialize(d = false)
      @debug = d
      @ultimaSorpresa = Sorpresa.new
      init()
      if(@debug)
        Diario.instance.ocurre_evento("Debug mode: " + @debug.to_s)
      end
    end
    
    def alMazo(s)
      
      if(!@barajada)
        @sorpresas.push(s)
      end
    end
    
    def siguiente()
      if((!@barajada || @usadas == @sorpresas.length) && !@debug)
        @sorpresas.shuffle
        @barajada = true
        @usadas = 0
        @usadas += 1
        temp = @sorpresas[0]
        @sorpresas.delete_at(0)
        @sorpresas.push(temp)
        @ultimaSorpresa = temp
      end
      
      return @ultimaSorpresa
    end
    
    def inhabilitarCartaEspecial(sorpresa)
      if(@sorpresas.include?(sorpresa))
        @sorpresas.delete(sorpresa)
        @cartasEspeciales.push(sorpresa)
        
        Diario.instance.ocurre_evento("Carta aniadida a cartas especiales")
      end
    end
    
    def habilitarCartaEspecial(sorpresa)
      if(@cartasEspeciales.include?(sorpresa))
        @cartasEspeciales.delete(sorpresa)
        @sorpresas.push(sorpresa)
        Diario.instance.ocurre_evento("Carta aniadida a sorpresas")
      end
    end
    
    private
    def init()
      @sorpresas = []
      @cartasEspeciales = []
      @barajada = false
      @usadas = 0
    end
  end
end
